### PR TITLE
stable diffusion model

### PR DESCRIPTION
**Pull request by MarcelWilnicki**
this PR adds a stable diffusion model.

prerequisites:

1. stable diffusion submodule must be checked out and pythonpath must point to it.
2. download model from here: https://ampereaimodelzoo.s3.eu-central-1.amazonaws.com/v2-1_512-ema-pruned.ckpt
3. script must point to the model with -m flag

**python packages:**
kornia
omegaconf
imwatermark
einops==0.6.0
open_clip_torch
invisible-watermark
pytorch_lightning==1.6.5

**example command:**
`AIO_NUM_THREADS=128 OMP_NUM_THREADS=128 numactl --membind=0 --cpubind=0 python aml/text_to_image/stable_diffusion/run.py -m /ampere/v2-1_512-ema-pruned.ckpt --num_runs 3 -b 3`

### CODE REVIEW

**Comment by jan-grzybek-ampere** [comment]
Which of the run*.py files should I review for now?

**Comment by MarcelWilnicki** [comment]
> Which of the run*.py files should I review for now?

sorry, for the confusion, I will need to clean it up, run2.py is fine

File: setup_deb.sh [hunk 1]
```diff
@@ -138,7 +138,23 @@ pip3 install --no-deps --upgrade \
...
+   imageio-ffmpeg==0.4.2 \
+   pytorch-lightning==1.4.2 \
+   torchmetrics==0.6 \
+   omegaconf==2.1.1 \
+   test-tube>=0.7.5 \
+   streamlit>=0.73.1 \
+   einops==0.3.0 \
+   webdataset==0.2.5 \
+   open-clip-torch==2.7.0 \
+   gradio==3.13.2 \
```
[the remark this answers is not in the document]
Comment (jan-grzybek-ampere) [reply setup_deb.sh:154]: Various other issues as well. When testing, run a clean release docker or even better a clean ubuntu 22 docker and make sure setup_deb.sh does all that's needed.

File: set_env_variables.sh [hunk 2]
```diff
@@ -39,5 +39,6 @@ if [ "${ARCH}" = "aarch64" ]; then
 fi
 export LD_PRELOAD=$LD_PRELOAD
 export PYTHONPATH=$SCRIPT_DIR
+export PYTHONPATH=$SCRIPT_DIR:$SCRIPT_DIR/text_to_image/stable_diffusion/stablediffusion
```
Comment (jan-grzybek-ampere) [inline set_env_variables.sh:42]: Unnecessary.

File: text_to_image/stable_diffusion/run.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+if __name__ == "__main__":
+    from utils.helpers import DefaultArgParser
+    parser = DefaultArgParser(["pytorch"])
+    parser.ask_for_batch_size()
+    parser.require_model_path()
+    parser.add_argument("--config", type=str,
+                        default="stablediffusion/configs/stable-diffusion/intel/v2-inference-fp32.yaml",
+                        help="path to config which constructs model")
+    parser.add_argument("--steps", type=int, default=25, help="steps through which the model processes the input")
+    parser.add_argument('--scale', type=int, default=9, help="scale of the image")
```
Comment (jan-grzybek-ampere) [inline text_to_image/stable_diffusion/run.py:91]: How this scale works?

File: text_to_image/stable_diffusion/run.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        else:
+            samples_ddim = torch.ones(1, 4, 96, 96, dtype=torch.float32)
+            scripted_decoder = torch.jit.trace(decoder, (samples_ddim))
+            scripted_decoder = torch.jit.freeze(scripted_decoder)
+            torch.jit.save(scripted_decoder, decoder_path)
+        model.first_stage_model.decoder = scripted_decoder
+
+    def single_pass_pytorch(_runner, _stablediffusion):
+        _runner.run(batch_size * steps)
+        _stablediffusion.submit_count()
```
Comment (jan-grzybek-ampere) [inline text_to_image/stable_diffusion/run.py:61]: it would be better to submit count based on batch size to know how many images have been generated (even though we dont measure accuracy yet)

File: text_to_image/stable_diffusion/run.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                                        conditioning=model.get_learned_conditioning([prompt] * batch_size),
+                                        batch_size=batch_size,
+                                        shape=shape,
+                                        verbose=False,
+                                        unconditional_guidance_scale=scale,
+                                        unconditional_conditioning=uc,
+                                        eta=0.0,
+                                        x_T=None)
+
+    runner = PyTorchRunnerV2(wrapper)
```
Comment (jan-grzybek-ampere) [inline text_to_image/stable_diffusion/run.py:76]: You could pass sampler.sample func here (without executing it) instead of wrapper. Then pass all variables (S=steps, conditioning, ...) in single_pass_pytorch func after task_size, i.e.: _runner.run(batch_size * steps, S=steps, conditioning=...., ....)

Comment (jan-grzybek-ampere) [reply text_to_image/stable_diffusion/run.py:76]: What about:

```
x_samples = model.decode_first_stage(samples)
x_samples = torch.clamp((x_samples + 1.0) / 2.0, min=0.0, max=1.0)

 for x_sample in x_samples:
    x_sample = 255. * rearrange(x_sample.cpu().numpy(), 'c h w -> h w c')
    img = Image.fromarray(x_sample.astype(np.uint8))
```

Full workload is from text to image.

**Review by jan-grzybek-ampere — CHANGES_REQUESTED** [review]
Nice PR. Obviously works and is nicely structured.

What remains:
- read-me, how to run this
- fixes to all the comments
- accuracy mode? https://huggingface.co/docs/diffusers/conceptual/evaluation - I suggest polishing this PR to make it mergeable, merging it, and then working on separate branch on this - you can also branch out right now and start working on this and then keep your separate accuracy branch updated with this one
- model zoo dev part

File: text_to_image/stable_diffusion/run.py (hunk 3 above)
Comment (jan-grzybek-ampere) [reply text_to_image/stable_diffusion/run.py:91]: I mean for my understanding only - not that there's something wrong about this.

**Comment by jan-grzybek-ampere** [comment]
Also you have a conflict with main branch now in setup_deb.sh (minor stuff)

File: utils/text_to_image/stable_diffusion.py [hunk 6]
```diff
@@ -8,7 +10,12 @@ def __init__(self):
         self.available_instances = 100000
 
     def get_input(self):
-        return 'a professional photograph of an astronaut riding a triceratops'
+        adjectives = ['professional', 'enormous', 'fun', 'beautiful', 'small', 'ugly']
+        nouns = ['dog', 'cat', 'astronaut', 'person', 'knight', 'horse', 'soldier']
+        actions = ['runs', 'jumps', 'rides a triceratops', 'rides a bike', 'eats a burger', 'washes clothes']
+        adverbs = ['quickly', 'slowly', 'loudly']
+
+        return f'{random.choice(adjectives)} {random.choice(nouns)} {random.choice(actions)} {random.choice(adverbs)}'
```
Comment (jan-grzybek-ampere) [inline utils/text_to_image/stable_diffusion.py:18]: Cool idea. You could also set random seed to make sure we always get the same combinations.

File: text_to_image/stable_diffusion/stablediffusion
[the remark this answers is not in the document]
Comment (MarcelWilnicki) [reply text_to_image/stable_diffusion/stablediffusion:1]: yes, I had to make some changes to the repo, and I am using a different commit. the PR is here: 
https://github.com/AmpereComputingAI/stablediffusion/pull/2

File: text_to_image/stable_diffusion/run.py (hunk 5 above)
Comment (MarcelWilnicki) [reply text_to_image/stable_diffusion/run.py]: in you second message do you suggest to save images? I don't do that at the moment because I don't measure the accuracy and thus I think it would be irrelevant for this PR.

**Review by jan-grzybek-ampere — CHANGES_REQUESTED** [review]
Last one thing is to make the benchmarking function benchmark whole pipeline - i.e. make runner.run() return ready images (in form of numpy arrays) that are then given to submit() function of the dataset object (even though you wont do anything with these images yet)

**Review by jan-grzybek-ampere — APPROVED** [review]
Looks good! Let's just resolve merge conflicts. Please test on clean release torch docker if everything (setup & runtime) works before merging.
